RecoLuminosity/LumiProducer/plugins/LumiProducerFromBrilcalc.cc

```diff
@@ -0,0 +1,217 @@
+  }
+  edm::LogInfo("LumiProducerFromBrilcalc") << "Read " << nLS << " lumisections from " << lumiFile_;
+  lumiFile.close();
+}
+
+LumiProducerFromBrilcalc::~LumiProducerFromBrilcalc() {
+  // do anything here that needs to be done at destruction time
+  // (e.g. close files, deallocate resources etc.)
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void LumiProducerFromBrilcalc::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  std::vector<float> bxlumi(3564, 0);
+  std::pair<int, int> runls = std::make_pair(iEvent.run(), iEvent.luminosityBlock());
+  if (lumiData_.count(runls) == 1) {
+    // if we have data for this run/LS, put it in the event
+    LogDebug("LumiProducerFromBrilcalc") << "Filling for run " << runls.first << " ls " << runls.second
+                                         << " with delivered " << lumiData_[runls].first << " dt "
+                                         << lumiData_[runls].second;
+    iEvent.put(std::make_unique<LumiInfo>(lumiData_[runls].second, bxlumi, lumiData_[runls].first), "brilcalc");
+  } else {
+    if (throwIfNotFound_) {
+      throw cms::Exception("LumiProducerFromBrilcalc")
+          << "Failed to find luminosity for run " << runls.first << " LS " << runls.second;
+    } else {
+      // just put in zeroes
+      edm::LogWarning("LumiProducerFromBrilcalc")
+          << "Failed to find luminosity for run " << runls.first << " ls " << runls.second;
+      iEvent.put(std::make_unique<LumiInfo>(0, bxlumi, 0), "brilcalc");
+    }
+  }
+}
+
+// ------------ method called once each stream before processing any runs, lumis or events  ------------
+void LumiProducerFromBrilcalc::beginStream(edm::StreamID) {}
+
+// ------------ method called once each stream after processing all runs, lumis and events  ------------
+void LumiProducerFromBrilcalc::endStream() {}
+
+// ------------ method called when starting to processes a run  ------------
+void LumiProducerFromBrilcalc::beginRun(edm::Run const&, edm::EventSetup const&) {}
+
+// ------------ method called when ending the processing of a run  ------------
+/*
+  void LumiProducerFromBrilcalc::endRun(edm::Run const&, edm::EventSetup const&) {
+  }
+*/
+
+// ------------ method called when starting to processes a luminosity block  ------------
+/*
+  void LumiProducerFromBrilcalc::beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) {
+  }
+*/
+
+// ------------ method called when ending the processing of a luminosity block  ------------
+/*
+  void LumiProducerFromBrilcalc::endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) {
+  }
+*/
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void LumiProducerFromBrilcalc::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // Allowed parameters
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<std::string>("lumiFile");
```

Comment:
Ok, this is now used only in private analyses, therefore it is mostly a matter of taste whether to define a parameter tracked or untracked.
On the other hand, if this is going to be processed centrally, it will certainly become important to keep track of the lumi file used, or whether the missing runs in it had been given lumi=0 or not.
I think that all these parameters should be tracked, instead

Reply:
Right, sounds reasonable, changed.